src/MaxText/input_pipeline/_grain_data_processing.py

```diff
@@ -66,7 +68,31 @@ def create_dataset_from_pattern(pattern):
       source = grain.ArrayRecordDataSource(files)
       return grain.MapDataset.source(source)
 
-    if ";" in data_file_pattern:
+    # Handle mixture config with named datasets, allows flexibility in recovering checkpoints
```

Comment:
do you add mp_prefetch and packing after mixing?

if so there's going to be a small adjustment on the boundary of a packed batch